src/main/org/deidentifier/arx/ShadowModelMembershipRisk.java

```diff
@@ -336,30 +376,627 @@ private Data getCopy(DataHandle handle, int[] sample) {
+        }
+        
+        /**
+         * Extracts a column from the handle
+         * @param handle
+         * @param column
+         * @return
+         */
+        private String[] getColumnAsString(int column) {
+            String[] result = new String[handle.getNumRows()];
+            for (int row = 0; row < handle.getNumRows(); row++) {
+                result[row] = handle.getValue(row, column);
+            }
+            return result;
+        }
+        
+        /**
+         * Extracts a column from the handle
+         * @param handle
+         * @param column
+         * @return
+         */
+        private Double[] getColumnAsDouble(int column) {
+            Double[] result = new Double[handle.getNumRows()];
+            for (int row = 0; row < handle.getNumRows(); row++) {
+                try {
+                    result[row] = handle.getDouble(row, column);
+                } catch (ParseException e) {
+                    throw new RuntimeException("Error reading double column");
+                }
+            }
+            return result;
+        }
+        
+        /**
+         * Gets int representation for string label
+         * 
+         * @param str
+         * @return
+         */
+        Integer getCode(int row, int col) {
+            return handle.internalGetEncodedValue(row, col, false);  
+        }
+
+        /**
+         * Gets int representation for multiple string labels
+         * 
+         * @param str
+         * @ret
+         */
+        List<Integer> getCode(int col) {
+            List<Integer> result = new ArrayList<>();
+            for(int row = 0; row < handle.getNumRows(); row++) {
+                result.add(getCode(row, col));
+            }
+            return result;
+        }
+
+        /**
+         * Transforms array of arrays to flatten array
+         * 
+         * @param input
+         * @return
+         */
+        private double[] flattenArray(double[][] input) {
+            
+            // calculate size of flatten array
+            int outputLength = 0;
+            for(double[] part : input) {
+                outputLength += part.length;
+            }
+            
+            // copy into flatten array
+            double[] output = new double[outputLength];
+            int posOutput = 0;
+            for(double[] part : input) {
+                for(double value : part) {
+                    output[posOutput++] = value;
+                }
+            }
+            return output;
+        }
+        
+    }
+    
+    /**
+     * MetaData Object that is used to calculate and encapsul information of the reference population.
+     * 
+     * 
+     * @author Thierry
+     *
+     */
+    class MetaData {
+
+        /** DataHandle */
+        private DataHandle handle;
+
+        /** Columns to consider */
+        private int[]      columns;
+
+        /** Available Values */
+        private int[][]    availableValues;
```

Comment:
Most important feedback: This will not work, as the same values will have different codes (ints) assigned in different versions of the dataset (i.e. when different instances of Data are creates, as is the case in getCopy() which is called by getAnonymizedOutput()).